test/tests/functional/pbs_qrun.py

```diff
@@ -92,3 +94,44 @@ def test_invalid_host_val(self):
         msg = "Server is not up"
         self.assertTrue(self.server.isUp(), msg)
         self.logger.info("As expected server is up and running")
+
+    def test_qrun_hangs(self):
+        """
+        This test submit 500 jobs with differnt equivalence class,
+        turn of scheduling and qrun job to
+        verify whether qrun hangs.
+        """
+        node = self.mom.shortname
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'False'})
+        self.server.manager(MGR_CMD_SET, NODE,
+                            {'resources_available.ncpus': 1}, id=node)
+        for walltime in range(1, 501):
+            j = Job(TEST_USER)
+            a = {'Resource_List.walltime': walltime}
+            j.set_attributes(a)
+            if walltime == 500:
+                jid = self.server.submit(j)
+            else:
+                self.server.submit(j)
+        self.logger.info("Submitted 500 jobs with different walltime")
+
+        now = int(time.time())
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'})
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'False'})
+        pid = os.fork()
+        if pid > 0:
+            try:
+                self.server.runjob(jobid=jid)
+            except PbsRunError as e:
+                self.logger.info("Runjob throws error: " + e.msg[0])
+        else:
+            out = self.scheduler.log_match(
+                "Starting Scheduling Cycle",
+                allmatch=True, interval=5, starttime=now, max_attempts=10)
+            if len(out) < 2:
+                os.kill(os.getppid(), signal.SIGTERM)
+                self.assertFalse(True,
+                                 "Qrun didn't start another sched cycle")
```

Comment:
you should probably do os.waitpid() and make sure the child process is finished before you return from this function

Reply:
As we are killing here the hung process (after checking the logs), so we don't need here waitpid.

Waitpid will wait for the process to time out (hung period), but to avoid that (as the PTL test case should not hung) we are forking another process.

Reply:
I was suggesting doing a waitpid() after the if statement, which means it would be after you have already sent the kill signal to the process. I just the function to reap the child process cleanly before exiting.

Reply:
If I use a waitpid after the kill then it will throw error, as that process is already killed.
    os.waitpid(hungpid, 0)
OSError: [Errno 10] No child processes